File: src/com/medium/CombinationSumIII_216.java
/**
 * 
 */
package com.medium;

import java.util.List;
import java.util.ArrayList;

/**
 * @author Chris Lin
 * @version 1.0
 */
public class CombinationSumIII_216 {
    // 2019年12月18日
    // Use DFS
    // Time Complexity: O(2^9)
    // Space Complexity:O(n)
    // Runtime: 1 ms, faster than 55.43%
    // Memory Usage: 34 MB, less than 8.00%
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> mResult = new ArrayList<>();
        dfs(k, n, 0, 1, mResult, new ArrayList<Integer>());
        return mResult;
    }

    private void dfs(int k, int target, int sum, int mStart, List<List<Integer>> mResult, List<Integer> mCur) {
        if (sum == target && k == mCur.size()) {
            mResult.add(new ArrayList<>(mCur));
        }
        if (sum > target) {
            return;
        }

        for (int i = mStart; i <= 9; i++) {
            mCur.add(i);
            dfs(k, target, sum + i, i + 1, mResult, mCur);
            mCur.remove(mCur.size() - 1);
        }
    }
}

File: Golang_src/medium/combinationSumIii_216.go
package medium

// Use backtracking:
// n: length of digits
// Time Complexity: O(P(9, k)) = O((9!*k)/(9-k)!)
// Space Complexity:O(K)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 1.91 MB, less than 70.19%
func combinationSum3(k int, n int) [][]int {
	upper := (1 + 9) * 10 / 2
	result := [][]int{}
	if k > upper {
		return result
	}

	keep := []int{}

	var dfs func(idx, remain int)
	dfs = func(idx, remain int) {
		if len(keep) == k && remain == 0 {
			tmp := make([]int, len(keep))
			copy(tmp, keep)
			result = append(result, tmp)
			return
		}

		if remain < 0 {
			return
		}

		for i := idx; i <= 9; i++ {
			keep = append(keep, i)
			dfs(i+1, remain-i)
			keep = keep[:len(keep)-1]
		}
	}

	dfs(1, n)

	return result
}
